cpp/src/arrow/dataset/projector.cc

```diff
@@ -83,9 +90,19 @@ Status RecordBatchProjector::SetInputSchema(std::shared_ptr<Schema> from,
 
   for (int i = 0; i < to_->num_fields(); ++i) {
     const auto& field = to_->field(i);
-    int matching_index = from_->GetFieldIndex(field->name());
+    FieldRef ref(field->name());
+    auto matches = ref.FindAll(*from_);
+
+    if (matches.empty()) {
+      // Mark column i as missing by setting missing_columns_[i]
+      // to a non-null placeholder.
+      RETURN_NOT_OK(
+          MakeArrayOfNull(pool, to_->field(i)->type(), 0, &missing_columns_[i]));
+      column_indices_[i] = kNoMatch;
+    } else {
+      RETURN_NOT_OK(ref.CheckNonMultiple(matches, *from_));
+      int matching_index = matches[0][0];
```

Comment:
Are we ignoring other potential matches here? `FindAll` could return multiple matches if there are columns with the same name, no?

Reply:
On line 103, I assert that there are not multiple matches

Reply:
Uh, looks like I forgot my glasses somewhere... where?

Reply:
`CheckNonMultiple` returns an error status if there are multiple matches